src/gui/PdfFloatingToolbox.cpp

```diff
@@ -0,0 +1,188 @@
+            // to make sure `pdfFloatingToolbox' is showing completely
+            auto view = self->pdfTextSelection->getPageView();
+            int pageWidth = view->getDisplayWidth();
+            int pageHeight = view->getDisplayHeight();
+
+            allocation->x = std::min(self->floatingToolboxX, pageWidth - allocation->width / 2);
+            allocation->y = std::min(self->floatingToolboxY, pageHeight - allocation->height / 2);
+        } else {
+            allocation->x = self->floatingToolboxX;
+            allocation->y = self->floatingToolboxY;
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+void PdfFloatingToolbox::postAction() {
+    this->pdfTextSelection->isFinished = true;
+    auto view = this->pdfTextSelection->getPageView();
+
+    this->hide();
+    this->pdfTextSelection->clearSelection();
+
+    delete this->pdfTextSelection;
+    this->pdfTextSelection = nullptr;
+    view->pdfTextSelection = nullptr;
+
+    view->rerenderPage();
+}
+
+void PdfFloatingToolbox::highlightCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->createStrokesForHighlight();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::copyTextCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->copyText();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::underlineCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->createStrokesForUnderline();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::strikethroughCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->createStrokesForStrikethrough();
+    pft->postAction();
+}
+
+void PdfFloatingToolbox::closeCb(GtkButton* button, PdfFloatingToolbox* pft) { pft->postAction(); }
+
+void PdfFloatingToolbox::show() {
+    gtk_widget_hide(this->floatingToolbox);  // force showing in new position
+    gtk_widget_show_all(this->floatingToolbox);
+}
+
+void PdfFloatingToolbox::setSelectType(PdfTextSelectType type) { this->selectType = type; }
+
+PdfTextSelectType PdfFloatingToolbox::getSelectType() { return this->selectType; }
+
+void PdfFloatingToolbox::switchSelectTypeCb(GtkButton* button, PdfFloatingToolbox* pft) {
+    pft->switchSelectType();
+    pft->pdfTextSelection->select();
+    pft->pdfTextSelection->getPageView()->rerenderPage();
+}
+
+void PdfFloatingToolbox::switchSelectType() {
+    if (this->selectType != PdfTextSelectType::SELECT_IN_AREA) {
+        this->selectType = PdfTextSelectType::SELECT_IN_AREA;
+    } else {
+        this->selectType = PdfTextSelectType::SELECT_HEAD_TAIL;
+    }
+}
+
+void PdfFloatingToolbox::copyText() {
+    GtkClipboard* clipboard = gtk_widget_get_clipboard(this->theMainWindow->getWindow(), GDK_SELECTION_CLIPBOARD);
+
+    const gchar* text = this->pdfTextSelection->getSelectedText().c_str();
+
+    gtk_clipboard_set_text(clipboard, text, -1);
+}
+
+void PdfFloatingToolbox::createStrokes(PdfMarkerStyle position, PdfMarkerStyle width, int markerOpacity) {
```

Comment:
It looks like highlighting/underlining is broken when input stabilization is on:
![segfault on getStroke: this=0x0 ](https://user-images.githubusercontent.com/46334387/131270987-d076fb9e-f333-4281-abcf-9a4e9312bdf2.png)

<details><summary>My input stabilization settings</summary>

![image](https://user-images.githubusercontent.com/46334387/131271039-ce73907f-54ef-4091-8c52-d384925bbffe.png)

</details>

Reply:
I rewrote the stroke creation logic, can you check again?